x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IlmHealthIndicatorService.java

```diff
@@ -79,39 +175,203 @@ public HealthIndicatorResult calculate(boolean verbose, int maxAffectedResources
+            })
+            .toList();
+    }
+
+    private static HealthIndicatorDetails createDetails(
+        boolean verbose,
+        IndexLifecycleMetadata metadata,
+        OperationMode mode,
+        List<IndexMetadata> stagnatingIndices
+    ) {
+        if (verbose == false) {
             return HealthIndicatorDetails.EMPTY;
         }
+
+        var details = new HashMap<String, Object>();
+
+        details.put("ilm_status", mode);
+        details.put("policies", metadata.getPolicies().size());
+        details.put("stagnating_indices", stagnatingIndices.size());
+
+        var stagnatingIndicesPerAction = stagnatingIndices.stream()
+            .collect(groupingBy(md -> md.getLifecycleExecutionState().action(), counting()));
+
+        if (stagnatingIndicesPerAction.isEmpty() == false) {
+            RULES_BY_ACTION_CONFIG.forEach((action, value) -> stagnatingIndicesPerAction.putIfAbsent(action, 0L));
+            details.put("stagnating_indices_per_action", stagnatingIndicesPerAction);
+        }
+
+        return new SimpleHealthIndicatorDetails(details);
+    }
+
+    /**
+     * Class in charge of find all the indices that are _potentially_ stagnated at some ILM action. To find the indices, it uses a list of
+     * rules evaluators (Check {@link org.elasticsearch.xpack.ilm.IlmHealthIndicatorService#RULES_BY_ACTION_CONFIG to the current rules}
+     */
+    static class StagnatingIndicesFinder {
+
+        private final ClusterService clusterService;
+        private final StagnatingIndicesRuleEvaluator stagnatingIndicesRuleEvaluator;
+        private final LongSupplier nowSupplier;
+
+        StagnatingIndicesFinder(
+            ClusterService clusterService,
+            StagnatingIndicesRuleEvaluator stagnatingIndicesRuleEvaluator,
+            LongSupplier nowSupplier
+        ) {
+            this.clusterService = clusterService;
+            this.stagnatingIndicesRuleEvaluator = stagnatingIndicesRuleEvaluator;
+            this.nowSupplier = nowSupplier;
+        }
+
+        /**
+         * @return A list containing the ILM managed indices that are stagnated in any ILM action/step.
+         */
+        public List<IndexMetadata> find() {
+            var metadata = clusterService.state().metadata();
+            var now = nowSupplier.getAsLong();
+            return metadata.indices()
+                .values()
+                .stream()
+                .filter(metadata::isIndexManagedByILM)
+                .filter(md -> stagnatingIndicesRuleEvaluator.isStagnated(now, md))
+                .toList();
+        }
+    }
+
+    record StagnatingIndicesRuleEvaluator(List<RuleConfig> rules) {
+        public boolean isStagnated(Long now, IndexMetadata indexMetadata) {
+            return rules.stream().anyMatch(r -> r.test(now, indexMetadata));
+        }
+    }
+
+    interface RuleConfig extends BiPredicate<Long, IndexMetadata> {
+        default TimeValue safelyGetTimeValue(Long now, Long currentTime) {
```

Comment:
This seems to be doiung more than just a conversion to `TimeValue`. Maybe `getElapsedTime` or something like that is a better name?